### DISCUSSION TITLE
Ash not loading some methods when loading Vulkan on Windows

### DISCUSSION THREAD

**Discussion by ChaseLewis** [ash-rs/ash#714] [Q&A] [answered]
I'm getting a panic when trying to create a swapchain.
 
 ``` 'Unable to load create_swapchain_khr' ```
 
 I'm following the tutorials at the bottom of the README.md but I had to modify them a bit to handle the new api for ash. Certain things seem to have changed since those were written. This panic seems to me it can't find a method called 'create_swapchain_khr' in the dll to call against.
 
 ```rs
 let entry = ash::Entry::linked()
 ```
 
 Main difference is linking to find Vulkan. I have VULKAN_SDK and VK_SDK_PATH variables. Vulkano doesn't have any issue working so don't think it has to do with my Vulkan installation at all. I'm using Vulkan 1.3.239, any ideas?

**Comment by MarijnS95** [comment, marked as answer]
> I'm following the tutorials at the bottom of the README.md but I had to modify them a bit to handle the new api for ash. Certain things seem to have changed since those were written.

As always, the repository for a crate may contain new development that has not made it into a public release yet. Either consult the corresponding `README.md` on crates.io or view the repo on the corresponding release tag, [likely `0.37.2` for the latest release](https://github.com/ash-rs/ash/tree/0.37.2).

(Related though, @Ralith it seems the removal of builder structs in #602 didn't make it to the readme _snippets_, should you or I correct that?)

> I'm getting a panic when trying to create a swapchain.
> 
> `'Unable to load create_swapchain_khr'`
> 
> This panic seems to me it can't find a method called 'create_swapchain_khr' in the dll to call against.

This means [`get_device_proc_addr` / `vkGetDeviceProcAddr`](https://github.com/ash-rs/ash/blob/7b3df8477d5de2422aa4a1718320ceb129769076/ash/src/extensions/khr/swapchain.rs#L20) returned `NULL` for the function in question (`vkCreateSwapchainKHR`):

https://github.com/ash-rs/ash/blob/7b3df8477d5de2422aa4a1718320ceb129769076/ash/src/vk/extensions.rs#L270-L279

(It doesn't need to be exported for that, depending on the implementation of the driver)
 
> ```rust
> let entry = ash::Entry::linked()
> ```
> 
> Main difference is linking to find Vulkan. I have VULKAN_SDK and VK_SDK_PATH variables. Vulkano doesn't have any issue working so don't think it has to do with my Vulkan installation at all. I'm using Vulkan 1.3.239, any ideas?

Linking or loading only affects how we get to `vkGetInstanceProcAddr`, the rest of the loading all happens through that function (and `vkGetDeviceProcAddr` which is loaded through it too). Rather, note that the swapchain is implemented as an extension which needs to be enabled first:

https://github.com/ash-rs/ash/blob/7b3df8477d5de2422aa4a1718320ceb129769076/examples/src/lib.rs#L319

If you enable the validation layer it _should™_ point this out.

**Reply by ChaseLewis** [reply]
ah I missed the extension being added to the logical device. 
Thanks.
